jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch

```diff
@@ -0,0 +1,116 @@
+From 8fe2aa05b85abe7aaf04625913bf6f2e987a3e49 Mon Sep 17 00:00:00 2001
+From: Dooyong Kim <kdooyong@amazon.com>
+Date: Tue, 24 Sep 2024 16:28:22 -0700
+Subject: [PATCH] tmp
+
+---
+ similarity_search/include/method/hnsw.h |  4 +++
+ similarity_search/src/method/hnsw.cc    | 36 +++++++++++++------------
+ 2 files changed, 23 insertions(+), 17 deletions(-)
+
+diff --git a/similarity_search/include/method/hnsw.h b/similarity_search/include/method/hnsw.h
+index 57d99d0..1174863 100644
+--- a/similarity_search/include/method/hnsw.h
++++ b/similarity_search/include/method/hnsw.h
+@@ -455,8 +455,12 @@ namespace similarity {
+     public:
+         virtual void SaveIndex(const string &location) override;
+ 
++        void SaveIndexWithStream(std::ostream& output);
++
+         virtual void LoadIndex(const string &location) override;
+ 
++        void LoadIndexWithStream(std::istream& input);
++
+         Hnsw(bool PrintProgress, const Space<dist_t> &space, const ObjectVector &data);
+         void CreateIndex(const AnyParams &IndexParams) override;
+ 
+diff --git a/similarity_search/src/method/hnsw.cc b/similarity_search/src/method/hnsw.cc
+index 35b372c..030fa75 100644
+--- a/similarity_search/src/method/hnsw.cc
++++ b/similarity_search/src/method/hnsw.cc
+@@ -746,13 +746,17 @@ namespace similarity {
+     }
+ 
+     template <typename dist_t>
+-    void
+-    Hnsw<dist_t>::SaveIndex(const string &location) {
+-        std::ofstream output(location,
+-                             std::ios::binary /* text files can be opened in binary mode as well */);
+-        CHECK_MSG(output, "Cannot open file '" + location + "' for writing");
+-        output.exceptions(ios::badbit | ios::failbit);
++    void Hnsw<dist_t>::SaveIndex(const string &location) {
```

Comment:
Can we just have SaveIndexWithStream parallel the SaveIndex constructor so we dont interfere too much with internals of methods?

Reply:
Got it. So leave the SaveIndex method as it is, and 'append' SaveIndexWithStream method to the class right?